engine/crates/engine-v2/query-planning/src/cost_estimation.rs

```diff
@@ -0,0 +1,156 @@
+use petgraph::{
+    graph::NodeIndex,
+    visit::{EdgeRef, IntoNodeReferences},
+    Direction,
+};
+use tracing::instrument;
+
+use crate::{Edge, Node, Operation, OperationGraph};
+
+const MAX_ITERATIONS: usize = 100;
+
+impl<'ctx, Op: Operation> OperationGraph<'ctx, Op> {
+    /// Gross approximation of the real cost of resolvers and fields. For any requirement we
+    /// compute the cost as the number of additional resolvers we need to go through.
+    ///
+    /// Supposing we the following fields: `{ a { b { c } e } }` and `e` depends on `c`. We'll roughly
+    /// generate the following operation graph: `a -> b -> c` and `a -> e`. We'll find their common
+    /// ancestor `a`, and from there the cost of `c` for `e` is the cost of `b` and `c`.
+    ///
+    /// As changing the cost of a resolver may impact any dependent ones, we need to iterate
+    /// several times to compute their real cost. If there is no cycles, which
+    /// composition should ensure, we will end up in a finite number of iterations until
+    /// convergence. But how many is unclear and a composition error shouldn't lead to an infinite
+    /// loop, so we're capping the max number of iterations to an arbitrary seemingly large enough
+    /// number.
+    ///
+    /// Today each iterations goes naively over all nodes again, we can certainly be smarter than
+    /// that.
+    #[instrument(skip_all)]
+    #[allow(unused)]
+    pub(crate) fn estimate_resolver_costs(&mut self) {
```

Comment:
Should we call this automatically when we create a new graph, or is it something the user must do when planning?

Reply:
We won't call this from the engine, it's separate so I can have a unit tests for it. Engine will just call a single function